futurex_openedx_extensions/helpers/roles.py

```diff
@@ -0,0 +1,346 @@
+from __future__ import annotations
+
+import logging
+from copy import deepcopy
+
+from common.djangoapps.student.models import CourseAccessRole
+from django.contrib.auth import get_user_model
+from django.db.models import OuterRef, Subquery
+from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+
+from futurex_openedx_extensions.helpers import constants as cs
+from futurex_openedx_extensions.helpers.caching import cache_dict
+from futurex_openedx_extensions.helpers.converters import error_details_to_dictionary, ids_string_to_list
+from futurex_openedx_extensions.helpers.models import ViewAllowedRoles
+from futurex_openedx_extensions.helpers.tenants import get_accessible_tenant_ids, get_all_tenant_ids
+
+logger = logging.getLogger(__name__)
+
+
+def is_valid_course_access_role(course_access_role: dict) -> bool:
+    """
+    Check if the course access role is valid.
+
+    :param course_access_role: The course access role
+    :type course_access_role: dict
+    :return: True if the course access role is valid, False otherwise
+    :rtype: bool
+    """
+    org = course_access_role['org'].strip()
+    course_id = course_access_role['course_id'] or ''
+
+    if not course_id and not org:
+        logger.error('Invalid course access role (both course_id and org are empty!): id=%s', course_access_role['id'])
+        return False
+
+    if course_id and not org:
+        logger.error('Invalid course access role (course_id with no org!): id=%s', course_access_role['id'])
+        return False
+
+    if course_id and org != course_access_role['course_org']:
+        logger.error('Invalid course access role (org mismatch!): id=%s', course_access_role['id'])
+        return False
+
+    return True
+
+
+def optimize_access_roles_result(access_roles: dict, course_org: dict):
+    """
+    Remove redundant course access roles that specify a course_id when the user already has access to the entire org.
+
+    :param access_roles: The access roles
+    :type access_roles: dict
+    :param course_org: The quick lookup for organization for each course
+    :type course_org: dict
+    """
+    for user_id, roles in access_roles.items():
+        for role, data in roles.items():
+            fixed_course_limited_access = []
+            orgs_of_courses = []
+            for course_id in data['course_limited_access']:
+                org = course_org[course_id]
+                if org not in data['orgs_full_access']:
+                    fixed_course_limited_access.append(course_id)
+                    if org not in orgs_of_courses:
+                        orgs_of_courses.append(org)
+            access_roles[user_id][role]['course_limited_access'] = fixed_course_limited_access
+            access_roles[user_id][role]['orgs_of_courses'] = orgs_of_courses
+
+
+@cache_dict(timeout='FX_CACHE_TIMEOUT_COURSE_ACCESS_ROLES', key_generator_or_name=cs.CACHE_NAME_ALL_COURSE_ACCESS_ROLES)
+def get_all_course_access_roles() -> dict:
```

Comment:
Note for Omar on performance: This is is probably an anti-pattern that slows down the request.

`#perf`